core/src/test/scala/org/apache/spark/SparkContextSuite.scala

```diff
@@ -1285,6 +1285,35 @@ class SparkContextSuite extends SparkFunSuite with LocalSparkContext with Eventu
       }
     }
   }
+
+  test("SPARK-35691: addFile/addJar/addDirectory should put CanonicalFile") {
+    withTempDir { dir =>
+      try {
+        sc = new SparkContext(new SparkConf().setAppName("test").setMaster("local"))
+
+        val sep = File.separator
+        val tmpCanonicalDir = Utils.createTempDir(dir.getAbsolutePath + sep + "test space")
+        val tmpAbsoluteDir = new File(tmpCanonicalDir.getAbsolutePath + sep + '.' + sep)
+        val tmpJar = File.createTempFile("test", ".jar", tmpAbsoluteDir)
+        val tmpFile = File.createTempFile("test", ".txt", tmpAbsoluteDir)
+
+        assert(tmpJar.getAbsolutePath != tmpJar.getCanonicalPath)
+        assert(tmpFile.getAbsolutePath != tmpFile.getCanonicalPath)
+
+        sc.addJar(tmpJar.getAbsolutePath)
+        sc.addFile(tmpFile.getAbsolutePath)
+
+        assert(sc.listJars().size == 1)
+        assert(sc.listFiles().size == 1)
```

Comment:
You can use the triple-equals here for Scalatest to give better error messages:
```
        assert(tmpJar.getAbsolutePath !== tmpJar.getCanonicalPath) // !== instead of !=
        assert(tmpFile.getAbsolutePath !== tmpFile.getCanonicalPath)
        ...
        assert(sc.listJars().size === 1)  // === instead of ==
        assert(sc.listFiles().size === 1)
```

Reply:
Thanks for the suggestion and review. :+1: